core nodes: Template node

ZJvandeWeg opened the pull request:
## Description

This core node use case doc is a bit light compared to others, as the main include docs from upstream are pretty good already.

## Related Issue(s)

<!-- What issue does this PR relate to? -->

## Checklist

<!-- https://flowforge.com/handbook/development/#defining-done -->

 - [ ] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [ ] I have considered the performance impact of these changes
 - [ ] Suitable unit/system level tests have been added and they pass
 - [ ] Documentation has been updated

ZJvandeWeg commented on src/_includes/core-nodes/template-use-case.md:
```diff
@@ -0,0 +1,27 @@
+
+For example, you can define a template like this: `<p>Hello {{payload.name}}!</p>`.
+
+![HTML Template with Mustache](./template-mustache.png "HTML template with Mustache")
```
```suggestion
![HTML Template with Mustache](./image/template-mustache.png "HTML template with Mustache")
```

Steve-Mcl commented on src/_includes/core-nodes/template-use-case.md:
````diff
@@ -0,0 +1,27 @@
+[{"id":"97d742eec1cb7dbf","type":"inject","z":"a6b7ede2e13fcbdf","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"{\"name\": \"FlowFuse\"}","payloadType":"json","x":180,"y":60,"wires":[["c85d70b41f374f02"]]},{"id":"c85d70b41f374f02","type":"template","z":"a6b7ede2e13fcbdf","name":"Template using payload.name", ...
+```
+
+This example will print `Hello FlowFuse!`.
````
```suggestion
This example will output `<p>Hello FlowFuse!</p>`.
```

Steve-Mcl reviewed (changes requested):
It reads fine and is accurate but I feel it could do with more examples and concentrate less on HTML.

For example, it is often used to make small JS objects from a JSON template with `{{mustache}}` values - and set to output parsed JSON. Similarly, it is used to create JS objects from YAML template and `{{mustache}}` values

Finally, sometimes, the template node is used for simply being a placeholder for data (no mustache). Users fill it with CSV or CSS or JSON etc, then use that to deliver content to a HTTP-Response or initiate some processing loop.  We dont mention that mode of operation either.

---


If this is not the focus, then apart from the 1 suggestion, I will approve.

If you are happy for me to do so, I would include a few more examples that folk could use as a reference.

Steve-Mcl commented:
@ZJvandeWeg 

2 examples I would propose we add:

### The Markdown
```md
    ## More examples

    ### Generating JavaScript objects

    #### input data
    ```js
    {
        payload: ['ACME', 1]
    }
    ```

    #### template node configuration
    * Template:
        ```text
        {
            "product": "{{payload.0}}",
            "version": {{payload.1}}
        }
        ```
    * Format: `Mustache template`
    * Output as: `Parsed JSON`

    #### output
    ```json
    { "product": "ACME", "version": 1 }
    ```

    ### Generating JavaScript object from YAML

    #### input data
    ```js
    {
        topic: 'FlowFuse',
        payload: 3000
    }
    ```

    #### template node configuration
    * Template:
        ```text
        options:
            title: {{topic}}
            port: {{payload}}
        ```
    * Format: `Mustache template`
    * Output as: `Parsed JSON`

    #### output
    ```json
    { "title": "FlowFuse", "port": 3000 }
    ```
```

<br>

### Renders as:
## More examples

### Generating JavaScript objects

#### input data
```js
{
    payload: ['ACME', 1]
}
```

#### template node configuration
* Template:
    ```text
    {
        "product": "{{payload.0}}",
        "version": {{payload.1}}
    }
    ```
* Format: `Mustache template`
* Output as: `Parsed JSON`

#### output
```json
{ "product": "ACME", "version": 1 }
```

### Generating JavaScript object from YAML

#### input data
```js
{
    topic: 'FlowFuse',
    payload: 3000
}
```

#### template node configuration
* Template:
    ```text
    options:
        title: {{topic}}
        port: {{payload}}
    ```
* Format: `Mustache template`
* Output as: `Parsed JSON`

#### output
```json
{ "title": "FlowFuse", "port": 3000 }
```

ZJvandeWeg commented:
@Steve-Mcl I've added your examples, needed to add the `{%- raw -%}` to make sure Nunjucks wasn't parsing the code examples. Further I've added some comments about comments, and about partials which aren't supported.

Steve-Mcl commented on src/_includes/core-nodes/template-use-case.md:
````diff
@@ -0,0 +1,116 @@
+In the Template node configuration, you define the template using text based formatting languages like HTML with the Mustache syntax.
+Mustache is a simple and "logic-less" templating language that allows you to insert variables and expressions into your template.
+
+For example, you can define a template like this: `<p>Hello {{payload.name}}!</p>`.
+
+![HTML Template with Mustache](./images/template-mustache.png "HTML template with Mustache")
+
+```json
+[{"id":"97d742eec1cb7dbf","type":"inject","z":"a6b7ede2e13fcbdf","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"{\"name\": \"FlowFuse\"}","payloadType":"json","x":180,"y":60,"wires":[["c85d70b41f374f02"]]},{"id":"c85d70b41f374f02","type":"template","z":"a6b7ede2e13fcbdf","name":"Template using payload.name", ...
+```
````
doesnt render correctly
![image](https://github.com/flowforge/website/assets/44235289/3b50f43e-5662-4245-858c-00942acd515a)

Steve-Mcl commented on src/_includes/core-nodes/template-use-case.md:
````diff
@@ -0,0 +1,116 @@
+In the Template node configuration, you define the template using text based formatting languages like HTML with the Mustache syntax.
+Mustache is a simple and "logic-less" templating language that allows you to insert variables and expressions into your template.
+
+For example, you can define a template like this: `<p>Hello {{payload.name}}!</p>`.
+
+![HTML Template with Mustache](./images/template-mustache.png "HTML template with Mustache")
+
+```json
+[{"id":"97d742eec1cb7dbf","type":"inject","z":"a6b7ede2e13fcbdf","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"{\"name\": \"FlowFuse\"}","payloadType":"json","x":180,"y":60,"wires":[["c85d70b41f374f02"]]},{"id":"c85d70b41f374f02","type":"template","z":"a6b7ede2e13fcbdf","name":"Template using payload.name", ...
+```
````
Unfortunately, still NG.

See here: https://deploy-preview-1096--flowforge-website.netlify.app/node-red/core-nodes/template/

I suspect the mustache in line 16 and the JSON flow both need some kind of escape to render the mustache.

![image](https://github.com/flowforge/website/assets/44235289/2f40f696-f324-48c5-9820-0a19c1d7a753)

Steve-Mcl commented:
@ZJvandeWeg unsure why you have merged this. As it stands, it is inaccurate. see comment  https://github.com/flowforge/website/pull/1096#discussion_r1321289074
